File: app/src/main/java/com/onemoresecret/crypto/EncryptedMessage.java
package com.onemoresecret.crypto;

import com.onemoresecret.OmsDataOutputStream;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPublicKey;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

public class EncryptedMessage {
    private final byte[] message;

    public EncryptedMessage(byte[] message,
                            RSAPublicKey rsaPublicKey,
                            int rsaTransformationIdx,
                            int aesKeyLength,
                            int aesTransformationIdx)
            throws NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException,
            BadPaddingException, InvalidKeyException, InvalidAlgorithmParameterException {

        this.message = MessageComposer.createRsaAesEnvelope(
                rsaPublicKey,
                rsaTransformationIdx,
                aesKeyLength,
                aesTransformationIdx,
                createPayload(getApplicationId(), message));
    }

    private byte[] createPayload(int ai, byte[] message) {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
             OmsDataOutputStream dataOutputStream = new OmsDataOutputStream(baos)) {

            // (1) the real Application Identifier
            dataOutputStream.writeUnsignedShort(ai);

            // (2) message key as byte array
            dataOutputStream.writeByteArray(message);

            return baos.toByteArray();
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }

    protected int getApplicationId() {
        return MessageComposer.APPLICATION_ENCRYPTED_MESSAGE;
    }

    public byte[] getMessage() {
        return message;
    }

}

File: app/src/main/java/com/onemoresecret/crypto/EncryptedMessage.kt
package com.onemoresecret.crypto

import com.onemoresecret.OmsDataOutputStream
import java.io.ByteArrayOutputStream
import java.io.IOException

open class EncryptedMessage(
    message: ByteArray,
    rsaPublicKeyMaterial: ByteArray,
    rsaTransformation: RsaTransformation,
    aesKeyLength: Int,
    aesTransformation: AesTransformation
) {
    @JvmField
    val message: ByteArray

    init {
        this.message = MessageComposer.createRsaAesEnvelope(
            rsaPublicKeyMaterial,
            rsaTransformation,
            aesKeyLength,
            aesTransformation,
            createPayload(this.applicationId, message)
        )
    }

    private fun createPayload(ai: Int, message: ByteArray): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                OmsDataOutputStream(baos).use { dataOutputStream ->

                    // (1) the real Application Identifier
                    dataOutputStream.writeUnsignedShort(ai)

                    // (2) message key as byte array
                    dataOutputStream.writeByteArray(message)
                    return baos.toByteArray()
                }
            }
        } catch (ex: IOException) {
            throw RuntimeException(ex)
        }
    }

    protected open val applicationId: Int
        get() = MessageComposer.APPLICATION_ENCRYPTED_MESSAGE
}
